synapseclient/models/team.py

```diff
@@ -1 +1,304 @@
+        self.modified_on = synapse_team.get("modifiedOn", None)
+        self.created_by = synapse_team.get("createdBy", None)
+        self.modified_by = synapse_team.get("modifiedBy", None)
+        return self
+
+    @otel_trace_method(
+        method_to_trace_name=lambda self, **kwargs: f"Team_Create: {self.name}"
+    )
+    async def create(self, synapse_client: Optional[Synapse] = None) -> "Team":
+        """Creates a new team on Synapse.
+
+        Args:
+            synapse_client: If not passed in or None this will use the last client from the `.login()` method.
+
+        Returns:
+            Team: The Team object.
+        """
+        loop = asyncio.get_event_loop()
+        current_context = context.get_current()
+        team = await loop.run_in_executor(
+            None,
+            lambda: Synapse.get_client(synapse_client=synapse_client).create_team(
+                name=self.name,
+                description=self.description if self.description else None,
+                icon=self.icon if self.icon else None,
+                opentelemetry_context=current_context,
+            ),
+        )
+        self.fill_from_dict(team)
+        return self
+
+    @otel_trace_method(
+        method_to_trace_name=lambda self, **kwargs: f"Team_Delete: {self.id}"
+    )
+    async def delete(self, synapse_client: Optional[Synapse] = None) -> None:
+        """Deletes a team from Synapse.
+
+        Args:
+            synapse_client: If not passed in or None this will use the last client from the `.login()` method.
+
+        Returns:
+            None
+        """
+        loop = asyncio.get_event_loop()
+        current_context = context.get_current()
+        delete = await loop.run_in_executor(
+            None,
+            lambda: Synapse.get_client(synapse_client=synapse_client).delete_team(
+                id=self.id, opentelemetry_context=current_context
+            ),
+        )
+        return delete
+
+    @otel_trace_method(
+        method_to_trace_name=lambda self, **kwargs: f"Team_From_Id: {self.id}"
+    )
+    async def from_id(
+        self, id: int, synapse_client: Optional[Synapse] = None
+    ) -> "Team":
+        """Gets Team object using its integer id.
+
+        Args:
+            id: The id of the team.
+            synapse_client: If not passed in or None this will use the last client from the `.login()` method.
+
+        Returns:
+            Team: The Team object.
+        """
+        loop = asyncio.get_event_loop()
+        current_context = context.get_current()
+        team = await loop.run_in_executor(
+            None,
+            lambda: Synapse.get_client(synapse_client=synapse_client).getTeam(
+                id=id, opentelemetry_context=current_context
+            ),
+        )
+        self.fill_from_dict(team)
+        return self
+
+    @otel_trace_method(
+        method_to_trace_name=lambda self, **kwargs: f"Team_From_Name: {self.name}"
+    )
+    async def from_name(
```

Comment:
I converted them both to classmethods in my latest commit

Reply:
I was thinking a bit about this - What are your thoughts around providing a classmethod AND a get method? My thought is around providing the most convenience for someone in these cases:

1. I have an instance of the class and I want to load it's current state from Synapse (ie: Call `get`)
2. I don't have an instance of the class but I want to create a loaded instance from Synapse (ie: Call `from_name` or `from_id`)

If we go down this idea path we could update the other models to follow this same idea - Where either:

1. `get` is a wrapper for these classmethods
2. OR the classmethods are wrappers for the `get`

Basically the 2 ideas being:
1.
```
@classmethod
  async def from_id(
      cls, id: int, synapse_client: Optional[Synapse] = None
  ) -> "Team":
    # Implementation

@classmethod
async def from_name(
    cls, name: str, synapse_client: Optional[Synapse] = None
) -> "Team":
    # Implementation

async def get(self, synapse_client: Optional[Synapse] = None) -> "Team":
    # Retrieve by ID `if self.id:` by calling `from_id`
    # Retrieve by Name `if self.name: by calling `from_name`
```
2.
```
@classmethod
  async def from_id(
      cls, id: int, synapse_client: Optional[Synapse] = None
  ) -> "Team":
    # Call GET

@classmethod
async def from_name(
    cls, name: str, synapse_client: Optional[Synapse] = None
) -> "Team":
    # Call GET

async def get(self, synapse_client: Optional[Synapse] = None) -> "Team":
    # Retrieve by ID `if self.id:` 
        # IMPLEMENTATION
    # Retrieve by Name `if self.name: 
        # IMPLEMENTATION
```

Whichever we want to follow we could also apply it to the other Models where we have similar rules to follow:
```
Project:
- Name
or
- ID

Folder:
- ID
or
- Name and Parent

File:
- ID
or
- Name and Parent
or
- MD5 (List of matching Files)

Table:
- ID
or
- Name and Project

Team:
- ID
or
- Name
```

cc @BWMac @thomasyu888 @jaymedina

Reply:
I was thinking about this the way that I thought about this [ticket](https://sagebionetworks.jira.com/browse/SYNPY-1186?jql=project%20in%20(SYNPY)%20AND%20assignee%3D%22Brad%20Macdonald%22%0A), where we determined that the best solution was to separate the concerns of using a username and using an ID number. I'm not strongly opposed to including a `get` method of some sort, but I think that both cases

> I have an instance of the class and I want to load it's current state from Synapse (ie: Call get)
and
> I don't have an instance of the class but I want to create a loaded instance from Synapse (ie: Call from_name or > 
> from_id)

are already covered by the `from_name` and `from_id` class methods:

```
# Load current state of my team
my_team = await Team.from_id(id=3492636)
print(my_team)

Team(id=3492636, name='python-client-test-team-1', description='testing OOP interface', icon=None, can_public_join=False, can_request_membership=True, etag='6639cfde-c5bb-4984-b51d-cd4cd2982469', created_on='2024-01-26T18:08:55.071Z', modified_on='2024-01-29T16:13:01.741Z', created_by='3458117', modified_by='3458117')

# Then, I go into the UI and change the ability for the public to join
# Reload current state of my team

await my_team.from_id(id=3492636)
print(my_team)

Team(id=3492636, name='python-client-test-team-1', description='testing OOP interface', icon=None, can_public_join=True, can_request_membership=True, etag='6639cfde-c5bb-4984-b51d-cd4cd2982469', created_on='2024-01-26T18:08:55.071Z', modified_on='2024-01-29T16:13:01.741Z', created_by='3458117', modified_by='3458117')
```

So I think the convenience is already there, and we currently don't need if/else logic to route the user's query to the appropriate class method.

Reply:
What you describing in this case case wouldn't work if I understand correctly:

```
# Load current state of my team
my_team = await Team.from_id(id=3492636)
print(my_team)

Team(id=3492636, name='python-client-test-team-1', description='testing OOP interface', icon=None, can_public_join=False, can_request_membership=True, etag='6639cfde-c5bb-4984-b51d-cd4cd2982469', created_on='2024-01-26T18:08:55.071Z', modified_on='2024-01-29T16:13:01.741Z', created_by='3458117', modified_by='3458117')

# Then, I go into the UI and change the ability for the public to join
# Reload current state of my team

await my_team.from_id(id=3492636)
print(my_team)

```

What would not work:
```
# This is the first instance - We create `my_team` from our classmethod
my_team = await Team.from_id(id=3492636)
print(my_team)

# A new instance of `my_team` is created, but thrown away because we are not assigning it to anything. The `my_team` class instance is NOT updated because we are not updating `self` in `from_id`
await my_team.from_id(id=3492636)
print(my_team)
```

Please let me know if my assumption here is incorrect (Because it'll save us a lot of headache if I am)

Reply:
Ah right, you are correct. It would have worked that way before I changed them to classmethods though.

Reply:
> was thinking about this the way that I thought about this [ticket](https://sagebionetworks.jira.com/browse/SYNPY-1186?jql=project%20in%20(SYNPY)%20AND%20assignee%3D%22Brad%20Macdonald%22%0A), where we determined that the best solution was to separate the concerns of using a username and using an ID number.

Something to mention about this ticket - In this ticket both the username and an ID number were being passed into the method in the same parameter. Meaning we didn't know which one they actually meant (ID or a username that happens to be a numerical number).

However - In this case they are separate fields on the Team (`id` and `name`). If we specify the order which we would use them in `get` I think that covers our use-case. We could say "We will retrieve the Team from Synapse by `id` or `name` in that order depending on which is filled in."

One other thing I was thinking about is whichever route we want to take (If we make these changes) that we should make sure that the object we call `get` with is also updated. That would mean the `get` could look like:
 ```
    async def get(self, synapse_client: Optional[Synapse] = None) -> "Team":
        if self.id:
            new_instance = await Team.from_id(id=self.id, synapse_client=synapse_client)
            self.__dict__.update(new_instance.__dict__)
            return self
        elif self.name:
            new_instance = await Team.from_name(
                name=self.name, synapse_client=synapse_client
            )
            self.__dict__.update(new_instance.__dict__)
            return new_instance
        raise ValueError("Team must have either an id or a name")
```

OR if we go the other direction it would look something like:
```
    async def get(self, synapse_client: Optional[Synapse] = None) -> "Team":
        if self.id:
            loop = asyncio.get_event_loop()
            current_context = context.get_current()
            api_team = await loop.run_in_executor(
                None,
                lambda: Synapse.get_client(synapse_client=synapse_client).getTeam(
                    id=self.id, opentelemetry_context=current_context
                ),
            )
            return self.fill_from_dict(api_team)
        elif self.name:
            loop = asyncio.get_event_loop()
            current_context = context.get_current()
            api_team = await loop.run_in_executor(
                None,
                lambda: Synapse.get_client(synapse_client=synapse_client).getTeam(
                    id=self.name, opentelemetry_context=current_context
                ),
            )
            return self.fill_from_dict(api_team)
        raise ValueError("Team must have either an id or a name")

    @classmethod
    @otel_trace_method(
        method_to_trace_name=lambda cls, id, **kwargs: f"Team_From_Id: {id}"
    )
    async def from_id(cls, id: int, synapse_client: Optional[Synapse] = None) -> "Team":
        """Gets Team object using its integer id.

        Arguments:
            id: The id of the team.
            synapse_client: If not passed in or None this will use the last client from the `.login()` method.

        Returns:
            Team: The Team object.
        """

        return await cls(id=id).get(synapse_client=synapse_client)

    @classmethod
    @otel_trace_method(
        method_to_trace_name=lambda cls, name, **kwargs: f"Team_From_Name: {name}"
    )
    async def from_name(
        cls, name: str, synapse_client: Optional[Synapse] = None
    ) -> "Team":
        """Gets Team object using its string name.

        Arguments:
            name: The name of the team.
            synapse_client: If not passed in or None this will use the last client from the `.login()` method.

        Returns:
            Team: The Team object.
        """
        return await cls(name=name).get(synapse_client=synapse_client)
```

Reply:
I guess my gut feeling is to go with option 1, though I'm not sure I could make a strong case that it is better than 2. Do you have a preference @BryanFauble?

Reply:
I like the feeling of 2 better. Mainly because of this line:
`self.__dict__.update(new_instance.__dict__)`

That is required in the first implementation feels more like a hack than anything.